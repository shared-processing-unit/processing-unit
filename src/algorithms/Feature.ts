export const createFilter = (feature: Feature, splitOn: number) => {
    return new Set(feature.ref.slice(0, splitOn))
}

export const distinctYSize = ({ refY }: Feature) => {
    return new Set(refY).size
}

export const filterLeft = (feature: Feature, refToFilter: Set<number>) => {
    return applyFilter(feature, refToFilter, true)
}
export const filterRight = (feature: Feature, refToFilter: Set<number>) => {
    return applyFilter(feature, refToFilter, false)
}
const applyFilter = (
    { id, refY, ref }: Feature,
    refToFilter: Set<number>,
    left: boolean
) => {
    const filter = new Set(
        ref
            .map((value, key) => [key, value] as [number, number])
            .filter(([, value]) => refToFilter.has(value))
            .map(([key]) => key)
    )
    return {
        ref: ref.filter((_, key) =>
            left ? filter.has(key) : !filter.has(key)
        ),
        refY: refY.filter((_, key) =>
            left ? filter.has(key) : !filter.has(key)
        ),
        id,
    }
}

export default interface Feature {
    readonly ref: number[]
    readonly refY: number[]
    readonly id: number
}
