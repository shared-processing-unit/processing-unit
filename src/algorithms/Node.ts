export default class Node<T> {
    public readonly left?: Node<T>
    public readonly right?: Node<T>
    constructor(public readonly value: T, left?: Node<T>, right?: Node<T>) {
        left && (this.left = left)
        right && (this.right = right)
    }
}
