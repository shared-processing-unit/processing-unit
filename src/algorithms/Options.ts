export default class Options {
    minSamplesSplit: number = 1
    maxDepth?: number
}
