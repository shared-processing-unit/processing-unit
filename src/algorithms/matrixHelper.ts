export const reverse = <T>(matrix: T[]): T[] => {
    return [...matrix].reverse()
}
export const vectorAdd = (left: number[], right: number[]) => {
    return left.map((entry, index) => {
        return index === left.length - 1 ? entry : entry + right[index + 1]
    })
}

export const transpose = <T>(data: Array<T[]>) => {
    return data[0].map((_, i) => data.map(row => row[i]))
}
