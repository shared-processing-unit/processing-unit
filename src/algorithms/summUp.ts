export default (data: number[] | number[][]) => {
    const flatten = ([] as number[]).concat(...data) as number[]
    return flatten.reduce((a, b) => a + b, 0)
}
