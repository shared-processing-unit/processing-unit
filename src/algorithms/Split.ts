import Feature from './Feature'
import { reverse, vectorAdd } from './matrixHelper'

export const evaluate = <T>(data: T[]) => {
    const ginis = vectorAdd(
        weightedGinies(data),
        weightedGinies(reverse(data)).reverse()
    )
    const value: number = Math.max(...ginis)
    return { splitOn: ginis.indexOf(value), value }
}

const weightedGinies = <T>(data: T[]) => {
    const occurences = new Map<T, number>()
    return data.map((category, i) => {
        const occurence = occurences.get(category) || 0
        occurences.set(category, occurence + 1)
        return giniImpurity(Array.from(occurences.values())) / (i + 1)
    })
}
const giniImpurity = (occurence: number[]) => {
    return occurence.reduce((prev, cur) => (prev += cur * cur), 0)
}

export default class Split {
    public readonly featureIndex: number
    public readonly index1: number
    public readonly index2: number

    constructor(
        public readonly value: number,
        public readonly splitOn: number,
        feature: Feature
    ) {
        this.featureIndex = feature.featureId
        this.index1 = feature.indexes[splitOn]
        this.index2 = feature.indexes[splitOn + 1]
    }
}
