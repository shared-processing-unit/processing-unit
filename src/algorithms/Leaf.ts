import Split from './Split'
import Feature from './Feature'

export default class Leaf {
    public readonly category: number
    public readonly samples: Array<[number, number]>
    public readonly split?: Split

    constructor(feature: Feature, split?: Split) {
        const samples = this.getSamples(feature.refY)
        const category = this.getCategory(samples)
        this.category = category
        this.samples = samples
        split && (this.split = split)
    }
    private getSamples(data: number[]) {
        return Array.from(
            data.reduce((occurences, category) => {
                const occurence = occurences.get(category) || 0
                return occurences.set(category, occurence + 1)
            }, new Map<number, number>())
        )
    }
    private getCategory(value: [number, number][]) {
        const mostOccurentCategory = Math.max(...value.map(([, o]) => o))
        const [[category]] = value.filter(([, o]) => o === mostOccurentCategory)
        return category
    }
}
