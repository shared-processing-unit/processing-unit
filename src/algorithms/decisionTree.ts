import Split, { evaluate } from './Split'
import Leaf from './Leaf'
import Feature, { createFilter, filterLeft, filterRight } from './Feature'
import Node from './Node'
import Options from './Options'
import { transpose } from './matrixHelper'

export const decisionTree = (
    tensor: Feature[],
    options: Options
): Node<Leaf> => {
    if (
        tensor[0].refY.length <= options.minSamplesSplit ||
        new Set(tensor[0].refY).size === 1
    ) {
        return new Node(new Leaf(tensor[0]))
    }
    const split = bestSplit(tensor)
    const [left, right] = splitFeatures(tensor, split)
    return new Node(
        new Leaf(tensor[split.featureIndex], split),
        decisionTree(left, options),
        decisionTree(right, options)
    )
}

const splitFeatures = (tensor: Feature[], split: Split) => {
    const feature = tensor[split.featureIndex]
    const filter = createFilter(feature, split.splitOn + 1)
    return [
        tensor.map(f => filterLeft(f, filter)),
        tensor.map(f => filterRight(f, filter))
    ]
}

const bestSplit = (tensor: Feature[]) => {
    const splits = tensor.map(feature => {
        const ginis = evaluate(feature.refY)
        return Array.from(new Set(feature.value)).map(value => {
            const index = feature.value.indexOf(value)
            return new Split(ginis[index], index, feature)
        })
    })
    const evaluations = [].concat(...(splits as [])) as Split[]
    const ginis = evaluations.map(({ gini }) => gini)
    const bestGini = Math.max(...ginis)
    return evaluations[ginis.indexOf(bestGini)]
}

export const createRandomForest = (
    dataLength: number,
    nEstimators: number,
    sampleSize: number,
    filter: number[]
) => {
    return Array(nEstimators)
        .fill({})
        .map(() => {
            return Array(sampleSize)
                .fill({})
                .map(() => {
                    let randomValue = -1
                    do {
                        randomValue = Math.floor(Math.random() * dataLength)
                    } while (filter.indexOf(randomValue) !== -1)
                    return randomValue
                })
                .sort((a, b) => a - b)
        })
}

export const featureToString = (features: Feature[]) => {
    return features
        .map(feature => {
            return [
                feature.indexes.join(','),
                feature.value.join(','),
                feature.refY.join(',')
            ].join('\n')
        })
        .join('\n')
}

export const parseUnsortedCSV = (csv: string) => {
    const data = transpose(
        csv.split('\n').map(row => row.split(','))
    ).map(col =>
        col
            .map(value => Number.parseFloat(value))
            .map((value, refX) => ({ value, refX }))
    )
    const X = data.slice(0, -1)
    const [Y] = data.slice(-1)
    return X.map(col => {
        const sortedCol = col.sort((a, b) => a.value - b.value)

        const distinctValue = Array.from(
            new Set(sortedCol.map(cell => cell.value))
        )
        return [
            sortedCol.map(({ refX }) => refX).join(','),
            sortedCol
                .map(({ value }) => distinctValue.indexOf(value))
                .join(','),
            sortedCol.map(({ refX }) => Y[refX].value).join(',')
        ].join('\n')
    }).join('\n')
}

export const parseSortedCSV = (csv: string) => {
    const nofColumns = 3
    const data = csv
        .split('\n')
        .map(row => row.split(','))
        .map(col => col.map(value => Number.parseFloat(value)))

    return Array(data.length / nofColumns)
        .fill({})
        .map((_, featureId) => {
            return {
                featureId,
                indexes: data[featureId * nofColumns],
                value: data[featureId * nofColumns + 1],
                refY: data[featureId * nofColumns + 2]
            } as Feature
        })
}

export const createSamples = (features: Feature[], splitIndexes: number[]) => {
    const sample1 = features.map(({ featureId, indexes, value, refY }) => {
        return {
            featureId,
            indexes: indexes.filter((_, ii) => splitIndexes.indexOf(ii) === -1),
            value: value.filter((_, ii) => splitIndexes.indexOf(ii) === -1),
            refY: refY.filter((_, ii) => splitIndexes.indexOf(ii) === -1)
        } as Feature
    })
    const sample2 = features.map(({ featureId, indexes, value, refY }) => {
        return {
            featureId,
            indexes: indexes.filter((_, ii) => splitIndexes.indexOf(ii) !== -1),
            value: value.filter((_, ii) => splitIndexes.indexOf(ii) !== -1),
            refY: refY.filter((_, ii) => splitIndexes.indexOf(ii) !== -1)
        } as Feature
    })
    const valuesSample2 = splitIndexes.map(testIndex =>
        features.map(feature => {
            const index = feature.indexes.indexOf(testIndex)
            return feature.value[index]
        })
    )
    const expectedSample2 = splitIndexes.map(
        i => features[0].refY[features[0].indexes.indexOf(i)]
    )

    return { expectedSample2, valuesSample2, sample1, sample2 }
}
