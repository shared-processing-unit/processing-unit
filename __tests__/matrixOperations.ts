import { reverse, vectorAdd } from '../src/algorithms/matrixHelper'
describe('matrixOperations', () => {
    it('vectorAdd', () => {
        expect(
            vectorAdd([1, 2, 3, 4, 5], [-1, -2, -3, -4, -5])
        ).toMatchSnapshot()
    })

    it('reverse matrix should not change input', () => {
        const input = [
            [0, 0, 1, 0],
            [0, 1, 0, 0],
            [0, 0, 0, 1],
            [1, 0, 0, 0],
            [1, 0, 0, 0],
            [0, 0, 1, 0],
            [0, 1, 0, 0],
        ]

        reverse(input)
        expect(input).toMatchSnapshot()
    })
    it('reverse vector should not change input', () => {
        const input = [0, 0, 1, 0]
        reverse(input)
        expect(input).toMatchSnapshot()
    })
    it('reverse vector', () => {
        expect(reverse([0, 0, 1, 0])).toMatchSnapshot()
    })
})
