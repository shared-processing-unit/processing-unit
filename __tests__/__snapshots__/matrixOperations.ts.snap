// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matrixOperations reverse matrix should not change input 1`] = `
Array [
  Array [
    0,
    0,
    1,
    0,
  ],
  Array [
    0,
    1,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
  ],
  Array [
    1,
    0,
    0,
    0,
  ],
  Array [
    1,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    1,
    0,
  ],
  Array [
    0,
    1,
    0,
    0,
  ],
]
`;

exports[`matrixOperations reverse vector 1`] = `
Array [
  0,
  1,
  0,
  0,
]
`;

exports[`matrixOperations reverse vector should not change input 1`] = `
Array [
  0,
  0,
  1,
  0,
]
`;

exports[`matrixOperations vectorAdd 1`] = `
Array [
  -1,
  -1,
  -1,
  -1,
  5,
]
`;
